README.md
#Note: This repository was cloned from the website http://unhaut.x10host.com/psxsdk/

PSXSDK 0.6
ReadMe and FAQ
Last updated: June 3rd 2016
-----------------------------

What is PSXSDK?

PSXSDK is an unofficial and homebrewed software development kit for the Sony PlayStation(TM).
Its programming tools (compiler, assembler, linker, etc.) are based on GNU Binutils and GNU
Compiler Collection (GCC) while the PlayStation-specific tools, which you need to convert data
to PSX format, were either written from scratch or adapted from already existing open source programs.

It is expected that you are already familiar with programming. 
If you want to learn how to program, I recommend starting programming for the operating system 
your computer uses and not here.

This readme is structured as a FAQ which also contains information which is not exclusive to the PSXSDK,
but that is generally informative.

Q: What do I need to run my programs on my PlayStation console?

A: There are several methods to do that, either a modchip, a boot disk, a cheating device
like Caetla, or the "swap trick".

The "modchip" method involves soldering a chip on the motherboard of your PlayStation to make it
think it's reading an original licensed disc. The kind of modchip depends on the model of your console,
and they are quite hard to find nowadays (as of 2010). You might already have one, as it was very
common to install it in many parts of the world.

I might be actually telling you to ignore your laws (in case you have EUCD and DMCA), and I recommend
to do it. The constant eroding of our rights as citizens is getting very annoying, and these
laws need to be fought. Let's not be hypocrites.
Big money influences law (TM).

The "boot disk" method uses a specially-made CD-ROM which looks like it is an original licensed disc
to the PlayStation. Once booted they show you a menu where you can boot an unlicensed disc.
You might be able to still find one for sale.

The "cheating device" method was very used back in the golden days of the PlayStation scene.
I have never used this method so I cannot describe it here. Search for "caetla" on your favourite
search engine and something interesting about that should come up.

The "swap trick" consists in first putting an original licensed disc in the PlayStation, so that
the PlayStation can authenticate, then when the drive is still spinning you replace it with an
unlicensed disc, and the PlayStation boots the disc as it already authenticated earlier.
The exact method of the "swap trick" varies between different models of the console.
It ruins your CD-ROM drive in the long run and spare PSX CD-ROM drives aren't very easy to find
(maybe harder to find than entire consoles), so I recommend to use something else if you can.
Search for this on your favourite search engine, something will surely come up.

On SCPH1001 PlayStations (Early American models) there seems to be an easier way to do the swap trick, also called audio menu exploit.
Make the SCPH1001 PlayStation go into the BIOS menu by not inserting any CD,
then go to the cd player, let it read an original PS1 disc until it stops spinning,
swap the disc with a burned CD-R disc, and exit the cd player.

This trick might also work on early Japanese consoles but I'm not sure of that.

Q: How do I get started?

A: In most cases, you will simply need one of the precompiled toolchains tarballs which you can find
on the PSXSDK website (http://unhaut.fav.cc/psxsdk). Extract them so that the PSXSDK can
be found at /usr/local/psxsdk and you're ready.

If you want to do it the hard way, there is no precompiled tarball for your operating system or
you want to help in developing the PSXSDK, read below on how to build it from sources.

Q: How do I compile a program?

A: This is easy. It is done almost as you do it usually, but with an additional step:

psx-gcc -o myprogram myprogram.c
elf2exe myprogram myprogram.exe

Why the additional step? It is needed because the GNU compiler outputs an ELF executable file,
which can't be run on the PlayStation. elf2exe converts that ELF file to a PS-X EXE, which
is the PlayStation's executable format, and which can be booted.

Q: How do I make a CDROM image?

A: I will first describe how to make your executable boot.
There are two ways to make your executable boot: you either name it "PSX.EXE" and you
put it in the CDROM filesystem's root directory, or you set a file named "SYSTEM.CNF" up
which describes a bit of information about your executable.

A SYSTEM.CNF file looks like this:

BOOT = cdrom:PROGRAM.EXE;1
TCB = 4
EVENT = 16
STACK = 801FFFF0

The BOOT field contains the filename of your program (in this case PROGRAM.EXE in the
root directory of the CDROM filesystem). I would advise you to not touch the other fields.
Modifying STACK is even useless (as the starting address of the stack is reset by the PSXSDK
library to 0x801FFFF0)

Now, you can make the ISO9660 filesystem for the CDROM with any program which does that.
This, for example, is the command you would use to do it with the cdrtools:

mkisofs -o test.iso -V TEST -sysid PLAYSTATION cd_root

Then you will have to use mkpsxiso to convert the ISO image to a format the PlayStation understands
and to license it. mkpsxiso outputs a .BIN/.CUE pair, which is supported by most CD burning
applications.
NOTE: Some Linux distributions (like Debian and Mint) don't provide mkisofs, but genisoimage instead.
           Replace the command name in the command line and that's it.

mkpsxiso test.iso test.bin /usr/local/psxsdk/misc/infousa.dat

This will create test.bin and test.cue, and the license applied to the image will be American.

Now you can burn the .BIN/.CUE pair with your favourite burning application.

Q: Where can I find the function reference for the PSXSDK?

A: You can find it either in /usr/local/psxsdk/misc/docs or in the docs subdirectory in the source
code archive.

Q: Are there any examples which show how to use the PSXSDK library?

A: Yes. Download the examples archive from the PSXSDK website and compile the examples in it.
In this way you can also test that your PSXSDK compiler is working correctly.
You can build them all by running ``make build_examples''

Q: What does PSXSDK support?

A: It can use the GPU (or GS, Graphic Synthesizer), the SPU, handle the VBlank Interrupt, handle
the joysticks and read files from the CDROM drive. Which is more than enough to do a 2D game.

Q: What is missing from PSXSDK?

A: Geometry Transformation Engine (GTE) handling, root counter support and a lot more...
Even if the PlayStation was very common back in its golden days, there is scarce documentation about it.

Q: Why does the C library appear incomplete?

A: Many standard C library functions are wrapped around their implementation in the BIOS which keeps
code size low and eases implementation, but this also means that they have less features than
most C libraries of today. That isn't a big problem, because for games you don't need a lot of features.
Anyway, you can modify the PSXSDK to remove the wrapper and use your functions.
The PSXSDK already does that for BIOS functions which were found not working or too buggy (like memcpy).

Q: Help! My PlayStation crashed / something odd appeared on screen / other hardware bug

A: PSXSDK was primarily tested on a PAL SCPH-102 PSOne, on SCPH-5552, SCPH-7002 and SCPH-9002 PlayStations and on the PCSX emulator.
There are probably several hardware quirks and oddities I've not accounted for but which happen
on other models of the console. Please send an email if something like that happens.
Be sure to read realhw.txt in the documentation directory first, though, as it might be 
an easily fixable error on your part.

Q: What program can I use to convert my images to PlayStation format?

A: You can use bmp2tim which is included in the PSXSDK or any program which converts an image to
PlayStation TIM format. It is a fairly standard format in the PlayStation field.

Q: What program do I use in case I want to convert sound?

A: You can use the included wav2vag or any other program which converts to that format.
As it is the case for TIM, VAG is a quite standard format in the PlayStation environment.

VABs are just find compilations of more VAG files.

Q: Can I sell a game I make with my PSXSDK?

A: Yes, as you are not infringing any copyright.

Q: I want to use Caetla to upload and test programs on my PlayStation, but I cannot find a version of CatFlap for an operating system which is not DOS or Windows. Are there?

A: Very recently Fabio Battaglia (hkzlabs) released his CatFlap for Linux. You can use that on Linux, and use its source
code as a starting point for a port to other operating systems.

Q: When uploading and running programs with CatFlap, they hang after some PSXSDK initialization messages.
How can I make them not hang? I cannot run anything!

A: Insert a CD into the drive of your PlayStation. It doesn't have to be a PlayStation CD, any CD is good.
This happens due to the initialization routines of the CDROM subsystem. Hopefully it will be fixed someday.

Q: How can I change the writing and the logo at the "Licensed by Sony" PlayStation bootscreen?

A: That information is contained inside the license file that you use to license the CD-ROM image that you burn.
Thus, to modify the information you must use the "lictool" program included in the PSXSDK tools.
lictool modifies various information contained in the license file such as the writing, the logo, etc.
    
For example, to output a new modified American license file which uses your logo contained in new.tmd:
    
lictool /usr/local/psxsdk/share/licenses/infousa.dat my_new_license_file.dat -tmd=new.tmd
    
Remember to use my_new_license_file.dat when you use mkpsxiso, otherwise you will use the standard
American license file, and your new logo won't appear!
    
Logos are in TMD format, the format for 3D models in Sony's original development kit.
    
Q: My music module sounds wrong or weird when I play it with libmodplay

A: Most music module formats have a lot of built-in effects, which are often poorly documented,
very complex to implement and which are often thought for personal computer hardware which is less
limited than the PlayStation SPU in some aspects. Luckily most modules don't really use them
or sound quite right when those effects are not implemented.

Q: I heard that it is possible to make the PlayStation read CD-RW discs (CD-rewritable). Is this true?

A: Yes, it is. But it is only really usable if you want to do hardware programming tests, as the amount
that the CD drive of your PlayStation may be able to read might be around 128-150 kilobytes. You will
not be able to read original PlayStation games or CD-Rs, and if you try to do it, your laser pickup might
get stuck and you may have to unblock it by unscrewing it and sliding it with your hands...
With a relatively good laser and very good tuning you might be even able to play full fledged games.

To read CD-RWs on later model PlayStation/cd drives you have to tune the potentiometer on the CD drive
from its original position to 12-13 hours. It actually sort of looks like a clock. Remember the original
position to not ruin anything. Use a precision screwdriver to tune the potentiometer.
This will stress the laser in the CD drive more than it is in normal operation, and it will shorten its life.

The trick to tuning the potentiometer correctly is burning an audio track on the CD-RW and playing the audio track
in the PlayStation CD player. If the audio track plays smoothly enough, it will be able to read CD-RWs.

You may need several tries before you succeed booting a disk, this happens due to the extensive read checks done
by the PlayStation BIOS when booting a game.

This is my (Giuseppe Gatta's) experience, and it might work much better on your PlayStation/PS cd drive
or it might work much worse. PlayStation cd drives are very finicky, I myself have PlayStations which read
like new and some which almost do not read at all. The CD drive is the lowest quality item inside
a PlayStation.

On the other hand this is the cheapest way to test things on your modded PlayStation.

Q: I don't use double buffering in my program/game. Nothing is shown on the emulator but it displays fine
    on the real hardware. Where is the issue?
    
A: You are using a buggy GPU plugin. If you're using version 1.0.16 of the P.E.Op.S SoftX Driver or most other
P.E.Op.S GPU plugins, update to P.E.Op.S SoftX Driver 1.0.18 and the problem should go away.

Q: What is a .PSF music file?

A: A .PSF music file is in the Portable Sound Format which is used to store video game music data.
Actually the .PSF just contains a tiny bit of information data in its header and then a compressed executable
which can be executed by an emulator (or a video game console). A PSF player is just a stripped-down emulator,
which supports only the SPU and the executable inside the PSF is an executable hacked to play only music.
Originally it was meant only for the PlayStation but now there are music files in this platform-indipendent format
from many consoles. You can play a .psf music just by using your console or emulator. Use the psfex tool shipped
with the PSXSDK, extract the .exe from the PSF and then run the executable. 
More info here: http://www.neillcorlett.com/psf/

Q: Why are filenames limited to the MSDOS-like 8.3 naming convention?

A: Because the ISO-9660 filesystem used on PlayStation CDs has the 8.3 limit for filenames.
It is plain ISO-9660 without any extension like Joliet or Rock Ridge, which are commonly added on CDs meant
to be read on computers to overcome the filename limit and to add new features.
The PSXSDK can make use of TRANS.TBL files in the directories of your CD to emulate long filenames.
TRANS.TBL is a normal text file which assigns long filenames to normal ISO9660 filenames.
Its rows look like this:
F SYSTEM.CNF;1		system.cnf

What this row does is assign to long filename system.cnf the short filename SYSTEM.CNF;1
'F' denotes that system.cnf is a file, otherwise 'D' denotes directories.

Good ISO image making tools will support the automatic creation of a TRANS.TBL file for each directory at image creation.
For instance, adding the -T parameter to your mkisofs command line makes mkisofs do just that.

To make fopen() use long filenames you will have to use the cdromL: pseudo device instead of cdrom:
Spaces are allowed in paths and both the slash and the back slash can be used to separate directories.
Examples: cdromL:/Path To/My Filename , cdromL:\Path To\This Name

There is a perfomance penalty when using long filenames because doing that requires reading the
TRANS.TBL for every directory which is encountered.

firstfile() does not support long filenames (yet) and open() does not support them, either.

Q: I'm running <linux distribution> and audio doesn't work in PCSX!
A: Many new Linux distributions are using PulseAudio. Run 'padsp pcsx' instead of 'pcsx'.

Q: Why is this SDK called PSXSDK? The PSX is a DVR+PS2 hybrid released by Sony in Japan!
A: Almost nobody knows about the Japanese PSX from 2004, and on top of that, it was a marketing flop.
    Seriously, PSX ('PlayStation eXperimental') is the name the original PlayStation went by during      development and it continued being called like that later. The name stuck and it is here to stay.

Q: Can I use UPX to pack an executable I made with PSXSDK?
A: Sure, you can. Due to a technicality of elf2exe, you need to specify --force on the UPX command line.
    It seems to work well, but I did no extensive testing, so it is at your own risk.

Q: Why do programs not work on the ePSXe/pSX/... emulators?
A: Despite popular belief, ePSXe and pSX are not as accurate as it is thought.
    It is recommended to use PCSX Reloaded for testing.
    The rule of thumb is that when something works fine on real hardware, and the
    emulator freezes or does something unexpected, it's the emulator's fault.

Written by Giuseppe Gatta
E-mail: tails92@gmail.com


Changelog
2016-06-03 (0.6):
- size_t and ssize_t definitions moved from stdio.h to types.h, and types.h is included automatically by stdio.h
- Added strcoll() - as a wrapper to strcmp()
- File device enum is now public and moved to stdio.h 
  It was incorrectly private and in libpsx/src/libc.c before
- fopen(): path argument is now const char*
- rename() and remove(): both arguments are now const char*
- open(): filename argument is now const char*
- get_real_file_size() and get_file_size(): argument is now const char*
- firstfile(): name argument is now const char*
- Added stdin, stdout and stderr FILE structures
- Implemented fwrite(). Only works with stdout and stderr
- Implemented fputs(). Only works with stdout and stderr
- Implemented freopen()
- Implemented clearerr(), feof(), ferror() and fileno()
- Implemented placeholders for strerror() and strerror_r()
- Added libf3m, to playback S3M module files, adapted to PSXSDK from GreaseMonkey's F3M
- libm/k_rem_pio2.c: fixed misleading indentation at line 188.
  libm/e_asin.c: fixed misleading indentation at line 89.
  libm/w_exp.c: o_threshold and u_threshold variables only defined when
  _IEEE_LIBM is not defined
  Fixed warnings with GCC 6.1.0
- Compilation warnings fixed for some tools, and some cleanup in spasm/parser.c
- Fixed warnings with psxgte.h - specifier was put after and not before enumeration
- Fixed warnings in the examples

2015-07-29 (0.5.99):
- Added PSX_WarmReboot()
- Fixed extern directive for GsCurDrawEnvH: was wrongly specified for GsCurDrawEnvY
- NORMAL_LUMINOSITY is now an alias for NORMAL_LUMINANCE: "luminosity" is an incorrect term for the meaning
- Added PRFONT_RL(f), PRFONT_GL(f), RLFONT_BL(f) attributes to GsSetFontAttrib(). These allow to adjust the luminance factor for the color components used when drawing the font: in few words, when used correctly, the attributes can be used to change the color of the font on the fly, without using another color look-up table.
- Added set_cop_register(), get_cop_register() and run_cop_instruction().
  set_cop0_register() and get_cop0_register() are now aliases for set_cop_register() and
  get_cop_register() respectively.
- Added set_cop_ctrl_register() and get_cop_ctrl_register()
- Fixed some bugs in the Makefiles, thanks to rsoft and Missingno_force for providing patches
- Fixed a bug in bmp2tim where 4-bit images with a width not a multiply of 4, or 8-bit images with a width
  not a multiply of 8 were erronously generated and thus contained invalid data.
  It should not have been possible to generate such images in the first place.
- Fixed a bug in bmp2tim where unnecessary debug messages were printed on screen when converting an image.
- The path for objcopy hardcoded inside elf2exe is now taken from the OBJCOPY variable specified in Makefile.cfg,
  and not from the OBJCOPY_PATH define in the source code itself
- Added -fno-strict-overflow flag by default to Makefile.cfg and to the psx-gcc and psx-g++ scripts
- Multiplication registers HI and LO are now saved and restored by the vblank and root counter handlers -
  this could give problems if any division or multiplication instruction was used inside a handler.
  Fortunately no such instructions are used inside handlers most times.
- Greatly improved nv-SPASM - now it should be fully compatible (maybe save some quirks) 
  with the Hitmen assembler
- Added GsVPrintFont()

2014-04-22 (0.4.99):
- Fixed the Makefile for the tools, now it considers the HOST_* variables and removes executables
  correctly under Windows
- added --enable-languages=c to the command line to build GCC, in toolchain.txt
- Added abort() to libc
- Fixed Makefile for misc directory
- Fixed genscript.sh: generated link script playstation.x now correctly
  includes libgcc and does not rely on EXTERN directives
- New SsUpload() implementation contributed by Shendo, which contains workarounds for emulators. 
  Also works on ePSXe.
- Fixed wrong return value for McWriteSector()
- Introduced debug build, that can be built by specifying the PSXSDK_DEBUG
  flag at compile time - it has more verbose messages, diagnostics, etc.
- Very verbose messages are now only compiled in for the debug builds.
- Added support for the 1KB scratchpad, which is the data cache that went
  unused in the PlayStation - it acts as a kind of "fast RAM";
  it is accessible as the byte array __scratchpad
- Added RemoveRCntHandler()
- Removed the InitCARD() - StartCARD() - StopCARD() sequence in PSXSDK_Init() when using BIOS mode,
  so that the vblank handler would not crash - the fix was acknowledging the interrupt in psxsdk's internal handler
  (in start/start.s). Thanks to Shendo for reporting this.
- The shell to use when executing scripts can now be specified in Makefile.cfg by changing the HOST_SHELL variable
- Moved the prototypes for index(), rindex(), strcasecmp() and strncasecmp()
  from string.h to strings.h, to be more standard
- Implemented the ffs*() and the popcount*() family of functions
  (the prototypes are in string.h or strings.h)
- string.h now includes strings.h for backward compatibility - advice: do not rely on this.
  when you use something whose prototype is in strings.h, ALSO include strings.h!
- implemented bitstring.h, bit string manipulation macros
- implemented memccpy()
- index() and rindex() were wrongly implemented both in libpsx/src/libc/string.c and as
  macros in strings.h; now they are only implemented as macros in strings.h
- added swab()
- added defines for EXIT_SUCCESS and EXIT_FAILURE in stdlib.h
- __PSXSDK__ is now defined by psx-gcc, this makes easier for code
  to detect it is being compiled for the PSXSDK
- added defines for O_RDLOCK, O_WRLOCK, O_SCAN, O_RCOM, O_EXCL, O_ASYNC, O_NOWAIT 
  and O_NOBUF, and corrected define for O_APPEND (it is now correctly 256, not 8) in fcntl.h
- Added a very simple implementation of stat(). 
  Only the st_size, st_blksize and st_blocks fields are available in struct stat (sys/stat.h)

libmodplay/xm.c
/**
 * MODPlay
 *
 * XM (FastTracker 2.00) file format support
 *
 * by Giuseppe Gatta, 2013
 */

#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <stdlib.h>
#include "modplay.h"

// I love data alignment so much..

#define xmWord(addr)			__xmWord_(xmData, addr)
#define xmDWord(addr)			__xmDWord_(xmData, addr)

static unsigned short __xmWord_(unsigned char *c, unsigned int addr)
{
	return c[addr] | (c[addr+1]<<8);
}

static unsigned int __xmDWord_(unsigned char *c, unsigned int addr)
{
	return c[addr] | (c[addr+1]<<8) | (c[addr+2]<<16) | (c[addr+3]<<24);
}

//#define XMprintf	printf
#define XMprintf

static void *xmUpdateSpeed(ModMusic *m)
{
	m->divisions_sec = 24 * m->beats_minute;
	m->divisions_sec /= m->ticks_division;
	m->divisions_sec /= 60;
}

ModMusic *MODLoad_XM(void *d)
{
	unsigned char *xmData = d;
	ModMusic *music;
	int i, j, l, m;
	unsigned char c;
	unsigned short s;
	int totalSamples=0;

// Check ID text
	
	if(memcmp(&xmData[0], "Extended Module: ",
		17) != 0)
		return NULL;

// Tested only with version 0x0104 files.	
	
	if(xmWord(58) != 0x0104)
		return NULL;
	
// Allocate memory for ModPlay structure	
	music = malloc(sizeof(ModMusic));
	
	if(music == NULL)
		return NULL;

// Fill structure's title field
	
	memcpy(music->title, &xmData[17], 20);
	music->title[20] = 0;
	
// Ignore tracker name
	
/**** XM HEADER VARIABLES ****/
	unsigned int xmHeaderSize = xmDWord(60);
	unsigned short xmSongLength = xmWord(64);
	XMprintf("xmSongLength = %d\n", xmSongLength);
	unsigned short xmRestartPosition = xmWord(66);
	unsigned short xmNumOfChannels = xmWord(68);	
	unsigned short xmNumOfPatterns = xmWord(70);
	unsigned short xmNumOfInstruments = xmWord(72);
	unsigned short xmFlags = xmWord(74);
	unsigned short xmDefaultTempo = xmWord(76);
	unsigned short xmDefaultBPM = xmWord(78);
	unsigned char *xmPatternOrderTable = &xmData[80];
	
	music->song_pos_num = xmSongLength;
	music->channel_num = xmNumOfChannels;
	music->pattern_num = xmNumOfPatterns;
	music->beats_minute = xmDefaultBPM;
	music->ticks_division = xmDefaultTempo;
	music->xm_flags = xmFlags;
	
	xmUpdateSpeed(music);
	
	memcpy(music->pattern_tbl, xmPatternOrderTable, 256);
/***********************************/

	XMprintf("xmFlags = %d\n", xmFlags);
	
	// 336
	int curPtr = 336;
	
	int oldCurPtr = curPtr;
	int patternDataSize = 0;
		
// First get size of all pattern data
	for(i = 0; i < xmNumOfPatterns; i++)
	{
		unsigned int xmPatternHeaderLength = xmDWord(curPtr);
		unsigned short xmPatternNumOfRows = xmWord(curPtr+5);
		unsigned short xmPatternPackedSize = xmWord(curPtr+7);

		curPtr += xmPatternHeaderLength + xmPatternPackedSize;
		patternDataSize += (xmPatternNumOfRows * 5 * xmNumOfChannels); 
	}
		
	music->pattern_data = malloc(patternDataSize);
	
	curPtr = oldCurPtr;
	
	int patternDataPos = 0;
	
	for(i = 0; i < xmNumOfPatterns; i++)
	{
		XMprintf("Pattern %d\n", i);
		
		unsigned int xmPatternHeaderLength = xmDWord(curPtr);
		XMprintf("xmPatternHeaderLength = %d\n", xmPatternHeaderLength);
		unsigned char xmPackingType = xmData[curPtr+4];
		XMprintf("xmPackingType = %d\n", xmPackingType);
		unsigned short xmPatternNumOfRows = xmWord(curPtr+5);
		XMprintf("xmPatternNumOfRows = %d\n", xmPatternNumOfRows);
		unsigned short xmPatternPackedSize = xmWord(curPtr+7);
		XMprintf("xmPatternPackedSize = %d\n", xmPatternPackedSize);
		
		music->pattern_row_num[i] = xmPatternNumOfRows;
		
		curPtr += xmPatternHeaderLength;
		
		//printf("CURPTR= %d\n", curPtr);
		

				
		//curPtr = oldCurPtr;	
		oldCurPtr = curPtr;
				
		//printf("RESULT: %d\n", curPtr - oldCurPtr);	

				
		for(j = 0; j < xmPatternNumOfRows; j++)
		{
			for(l = 0; l < xmNumOfChannels; l++)
			{
				unsigned char rowNote = xmData[curPtr];
				unsigned char rowInstrument=0;
				unsigned char rowVolumeColumn=0;
				unsigned char rowEffectType=0;
				unsigned char rowEffectParameter=0;
			
				if(rowNote & 128) // Packing enabled
				{
					unsigned char packInfo = rowNote;
					//printf("Packing found.. %02x\n", packInfo);
				
					rowNote = 0;
					curPtr++;
				
					if(packInfo & 1) // Note follows
						rowNote = xmData[curPtr++];
					if(packInfo & 2) // Insturment follows
						rowInstrument = xmData[curPtr++];
					if(packInfo & 4)
						rowVolumeColumn = xmData[curPtr++];
					if(packInfo & 8)
						rowEffectType = xmData[curPtr++];
					if(packInfo & 16)
						rowEffectParameter = xmData[curPtr++];
				}
				else
				{
					rowInstrument = xmData[curPtr+1];
					rowVolumeColumn = xmData[curPtr+2];
					rowEffectType = xmData[curPtr+3];
					rowEffectParameter = xmData[curPtr+4];
					curPtr+=5;
				}

// The XM player code will always want unpacked pattern data.				
				music->pattern_data[patternDataPos++] = rowNote;
				music->pattern_data[patternDataPos++] = rowInstrument;
				music->pattern_data[patternDataPos++] = rowVolumeColumn;
				music->pattern_data[patternDataPos++] = rowEffectType;
				music->pattern_data[patternDataPos++] = rowEffectParameter;
			}
		}
			
			//printf("CURPTR=%d\n", curPtr);
		
		//curPtr += xmPatternHeaderLength + xmPatternPackedSize;
	}

	oldCurPtr = curPtr;
		
	for(l = 0; l < 2; l++)
	{
		// l = 0, PASS 1. Find number of samples in XM.
		// l = 1, PASS 2. Allocate & fill instrument and sample structures.
		
		
		curPtr = oldCurPtr;
		
		if(l == 1)
		{
			music->sample_num = totalSamples;
			music->instrument_num = xmNumOfInstruments;
			music->sample = malloc(sizeof(ModSample) * totalSamples);
			music->instrument = malloc(sizeof(ModInstrument) * xmNumOfInstruments);
		}
		
		int currentSample = 0;
		
		for(i = 0; i < xmNumOfInstruments; i++)
		{
			XMprintf("curPtrOff=%x\n", curPtr);

		//	printf("Instrument %d\n", i);
		
			unsigned int instrumentHeaderSize = xmDWord(curPtr);
			unsigned char *instrumentName = &xmData[curPtr+4];
		//	printf("instrumentName = %s\n", instrumentName);
			unsigned char instrumentType = xmData[curPtr+26];
			unsigned short instrumentNumOfSamples = xmWord(curPtr+27);
			
			XMprintf("instrumentNumOfSamples = %d\n", instrumentNumOfSamples);
			
			if(l == 0)
			{
				totalSamples += instrumentNumOfSamples;
				XMprintf("totalSamples=%d\n", totalSamples);
				XMprintf("%d: %s\n", i, instrumentName);
			}
			else
			{
				memcpy(music->instrument[i].name, instrumentName, 22);
				music->instrument[i].name[22] = 0;
				music->instrument[i].sample_num = instrumentNumOfSamples;
			}
			
		//printf("instrumentNumOfSamples = %d\n", instrumentNumOfSamples);
		
			unsigned int sampleHeaderSize = xmDWord(curPtr + 29);
			unsigned char *sampleNumberAllNotes = &xmData[curPtr + 33];
			unsigned char *samplePointsVolEnvelope = &xmData[curPtr + 129];
			unsigned char *samplePointsPanEnvelope = &xmData[curPtr + 177];
			unsigned char sampleNumOfVolumePoints = xmData[curPtr + 225];
			unsigned char sampleNumOfPanningPoints = xmData[curPtr + 226];
			unsigned char sampleVolumeSustainPoint = xmData[curPtr + 227];
			unsigned char sampleVolumeLoopStartPoint = xmData[curPtr + 228];
			unsigned char sampleVolumeLoopEndPoint = xmData[curPtr + 229];
			unsigned char samplePanningSustainPoint = xmData[curPtr + 230];
			unsigned char samplePanningLoopStartPoint = xmData[curPtr + 231];
			unsigned char samplePanningLoopEndPoint = xmData[curPtr + 232];
			unsigned char sampleVolumeType = xmData[curPtr + 233];
			unsigned char samplePanningType = xmData[curPtr + 234];
			unsigned char sampleVibratoType = xmData[curPtr + 235];
			unsigned char sampleVibratoSweep = xmData[curPtr + 236];
			unsigned char sampleVibratoDepth = xmData[curPtr + 237];
			unsigned char sampleVibratoRate = xmData[curPtr + 238];
			unsigned short sampleVolumeFadeout = xmDWord(curPtr+239);
		
			XMprintf("instrumentHeaderSize = %d\n", instrumentHeaderSize);
			curPtr += instrumentHeaderSize;
		
			for(j = 0; j < instrumentNumOfSamples; j++)
			{
				unsigned int sampleLength = xmDWord(curPtr);
				unsigned int sampleLoopStart = xmDWord(curPtr+4);
				unsigned int sampleLoopLength = xmDWord(curPtr+8);
				unsigned char sampleVolume = xmData[curPtr+12];
				unsigned char sampleFinetune = xmData[curPtr+13];
				unsigned char sampleType = xmData[curPtr+14];
				unsigned char samplePanning = xmData[curPtr+15];
				char sampleRelativeNoteNumber = xmData[curPtr+16];
				char sampleDataType = xmData[curPtr+17];
				
				XMprintf("sampleDataType = %02x, sampleLength=%d\n", sampleDataType,
					sampleLength);
				
				char *sampleName = &xmData[curPtr+18];
				XMprintf("sampleName[%d] = %s ( %s)\n", j, sampleName,
					(sampleType & 16)?"16-bit":"8-bit");
			
				char *sampleData = &xmData[curPtr + sampleHeaderSize];
			
				//curPtr+=sampleHeaderSize;
				
				if(l == 1)
				{
					memcpy(music->sample[currentSample].name, sampleName, 22);
					music->sample[currentSample].name[22] = 0;
					music->sample[currentSample].length = sampleLength;
					music->sample[currentSample].finetune = sampleFinetune;
					music->sample[currentSample].volume = sampleVolume;
					music->sample[currentSample].repeat_off = sampleLoopStart;
					music->sample[currentSample].repeat_len = sampleLoopLength;
					music->sample[currentSample].bits = (sampleType & 16) ? 16: 8;
					music->sample[currentSample].data_type = sampleDataType;
					music->sample[currentSample].data = malloc(sampleLength);
					
					/**** XM SPECIFIC ****/
					music->sample[currentSample].relative_note = sampleRelativeNoteNumber;
					
					switch(music->sample[currentSample].bits)
					{
						case 8:
							for(m = 0, c = 0; m < sampleLength; m++)
							{
								c += sampleData[m];
								music->sample[currentSample].data[m] = c ^ 0x80; 
							}
						break;
						case 16:
							for(m = 0, s = 0; m < sampleLength; m+=2)
							{
								s += (sampleData[m+1] << 8) | sampleData[m];
								//s ^= 0x8000;
								memcpy(&music->sample[currentSample].data[m], &s, sizeof(short));
							}
						break;
					}
					
					music->instrument[i].sample_ids[j] = currentSample;
					currentSample++;
				}
				
				curPtr += sampleHeaderSize + sampleLength;
			
				XMprintf("sampleHeaderSize = %d\n", sampleHeaderSize);
//			curPtr += sampleLength;
			
			/*unsigned char xmSampleNumber = xmData[curPtr + 33];
			unsigned char xmSamplePointsVolEnvelope = xmData[curPtr + 34];
			unsigned char xmSamplePointsPanEnvelope = xmData[cur*/
			//curPtr += sample
			}
		//curPtr += instrumentHeaderSize;
		}
	}
/***********************************/	
	
	music->fmt = MOD_FMT_XM;
	music->song_pos = 0;
	music->pat_pos = 0;
	
	return music;
}

void MODPlay_XM(ModMusic *m, int *t)
{
	int c, i;
	
	if(*t == 0)
		return;
	
	modplay_int_cnt++;

	if(modplay_int_cnt != (50 / m->divisions_sec))
		return;
	
	for(c = 0; c < m->channel_num; c++)
	{
		int curPat = m->pattern_tbl[m->song_pos];
		int curDataPos = 0;
		
// Find start of data for the current pattern		
		for(i = 0; i < curPat; i++)
			curDataPos += (m->pattern_row_num[i] * 5 * m->channel_num);
		
		

	}
}
